### ISSUE TITLE
Coherence between MontyEncoder/MontyDecoder and as_dict/from_dict

### ISSUE THREAD

**Issue by davidwaroquiers** [materialsvirtuallab/monty#44]
We have been thinking about extending the MontyEncoder and as_dict implementations such that they would be more "coherent" with the MontyDecoder and from_dict.

Currently, if you have a MSONable containing other MSONable classes, the standard monty implementation of as_dict and MontyEncoder do not go in the lower levels (e.g. of a list or a dict). This prevents from directly using the as_dict to write to a json file without reimplementing your own as_dict method. An example with a simple list is given below (similar problem occurs for dict and any nested lists/dicts).

The MontyDecoder and from_dict on the other hand, allows to reconstruct the object (if the standard MSONable's as_dict method of that object has been overwritten).

Before starting to work on this and make a solution (that would actually go through the list or dict and see if the object has an as_dict method, then use it), we wanted to know if :
1. You are not working on it
2. You have tried but had good reasons to abandon
3. You have not tried but had good reasons not to do it

Thank you for your comments,

David and Guido (@gpetretto)

## Example code

Here above an example that shows the issue.

```python
from monty.json import MSONable
import json


class ClassA(MSONable):
    def __init__(self, a, b, c):
        self.a = a
        self._b = b
        self.c = c


# Example class containing a dictionary whose values are ClassA objects (as_dict is explicitly implemented here and overwrites the one from MSONable)
class ClassB(MSONable):
    def __init__(self, class_a_list):
        assert isinstance(class_a_list, list)
        self.class_a_list = class_a_list

    def as_dict(self):
        return {"@module": self.__class__.__module__,
                    "@class": self.__class__.__name__,
                    "class_a_list": [a.as_dict() for a in self.class_a_list]}


# Example class containing a list of ClassA objects ("standard" as_dict inherited from MSONable)
class ClassC(MSONable):
    def __init__(self, class_a_list):
        assert isinstance(class_a_list, list)
        self.class_a_list = class_a_list


# Definition of 2 arbitrary ClassA objects
a1 = ClassA(a=1, b=2, c=3)
a2 = ClassA(a=4, b=5, c=6)


# Example when as_dict is reimplemented
class_b = ClassB(class_a_list=[a1, a2])
class_b_dict = class_b.as_dict()
class_b_json_str = json.dumps(class_b_dict)

# Example without reimplementing as_dict (using MSONable's one)
class_c = ClassC(class_a_list=[a1, a2])
class_c_dict = class_c.as_dict()
class_c_json_str = json.dumps(class_c_dict)  # This raises TypeError
```

## Error message

```
TypeError: Object of type 'ClassA' is not JSON serializable
```

## Suggested solution (if known)

* See above

**Comment by shyuep** [comment]
I have no objections. And I personally have not attempted any of the three bullet points. So go ahead and draft an implementation with appropriate unittests. Note that monty has to remain py27 and py3 compatibale for now.

**Comment by davidwaroquiers** [comment]
Solved in #45 .
